clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignmentReconciler.java

```diff
@@ -0,0 +1,263 @@
+
+/**
+ * {@code AssignmentReconciler} performs the work of reconciling this consumer's partition assignment as directed
+ * by the consumer group coordinator. When the coordinator determines that a change to the partition ownership of
+ * the group is required, it will communicate with each consumer to relay its respective <em>target</em>
+ * assignment, that is, the set of partitions for which that consumer should now assume ownership. It is the then the
+ * responsibility of the consumer to work toward that target by performing the necessary internal modifications to
+ * satisfy the assignment from the coordinator. In practical terms, this means that it must first determine the set
+ * difference between the <em>{@link SubscriptionState#assignedPartitions() current assignment}</em> and the
+ * <em>{@link Assignment#topicPartitions() target assignment}</em>.
+ *
+ * <p/>
+ *
+ * Internally, reconciliation requires the following steps:
+ *
+ * <ol>
+ *     <li>
+ *         On the background thread, upon receipt of a new assignment from the group coordinator, the
+ *         {@link MembershipManager} should call {@link #startReconcile(Assignment)} to start reconciliation.
+ *     </li>
+ *     <li>
+ *         Internally, the partitions to revoke are determined via {@link #getPartitionsToRevoke(Assignment)};
+ *         these are the partitions in the current assignment that are <em>not in</em> the target assignment.
+ *         Next, we calculate the partitions to assign using {@link #getPartitionsToAssign(Assignment)};
+ *         these are the partitions in the target assignment that are <em>not in</em> the current assignment.
+ *     </li>
+ *     <li>
+ *         Send a {@link PartitionReconciliationStartedEvent} so that the application thread will execute the
+ *         {@link ConsumerRebalanceListener} callback methods.
+ *     </li>
+ *     <li>
+ *         On the application thread, when the {@link PartitionReconciliationStartedEvent} is received, execute the
+ *         {@link ConsumerRebalanceListener#onPartitionsRevoked(Collection)} and
+ *         {@link ConsumerRebalanceListener#onPartitionsAssigned(Collection)} callback methods
+ *     </li>
+ *     <li>
+ *         Enqueue a corresponding {@link PartitionReconciliationCompleteEvent} so that the background thread
+ *         will know the listener was invoked and the result of the invocation
+ *     </li>
+ *     <li>
+ *         On the background thread, process the {@link PartitionReconciliationCompleteEvent}, which should call the
+ *         {@link MembershipManager#completeReconcile(Set, Set, Optional)} method. This method will call
+ *         {@link #completeReconcile(Set, Set)} to remove the revoked partitions and add the assigned partitions in the
+ *         {@link SubscriptionState#assignFromSubscribed(Collection) current assignment} and then make a note to
+ *         send a {@link ConsumerGroupHeartbeatRequest} to the group coordinator on its next pass of
```

Comment:
maybe just "and then rejoin by doing ####"

Reply:
Sorry, can you clarify where this comment should be placed? My understanding is that rejoining isn't necessary after reconciliation.